### ISSUE TITLE
Include support for MIPS64EL architecture

### ISSUE THREAD

**Issue by luisdanielsc** [BeyondTrust/pbis-open#178]
**Version**: 8.5.4.334
**OS/Distro**: Debian Stretch
**Issue/Impact**: All end users of Loongson 3 computers
I would like to know if it is possible to include support for the MIPS64EL architecture of the Loongson 3 computers.

**Comment by rbest-bt** [comment]
There are no current plans and we do not have the hardware.

**Comment by luisdanielsc** [comment]
> There are no current plans and we do not have the hardware.

Could you tell me how I can do it on my own? Is there a manual explaining how to add support for a new architecture in pbis-open?

**Comment by rbest-bt** [comment]
Follow the steps in the [Readme](https://github.com/BeyondTrust/pbis-open) and build it on the system in question. There maybe issues that will need to be worked out. If you manage to get it building/working you may want to look into [contributing](https://github.com/BeyondTrust/pbis-open/blob/master/CONTRIBUTING.md) back so that other maybe able to build on it.

**Comment by dmorash-BT** [comment]
Unfortunately this is no guide to porting to a new platform, but porting will likely involve modifying the following areas:

1. Update various MakeKit modules to support a new platform. 
-  makekit/mk/module/autotools.sh - update _mk_at_system_string() to correctly set the architecture for your platform
- makekit/mk/module/platform.sh - update option() to set the default MK_BUILD_OS and MK_BUILD_ARCH variables based on the platform PBIS is being built on
- makekit/mk/module/path.h - update option() to set the library directory, e.g. lib64
- makekit/mk/module/compiler.sh - update option() to set the default CFLAGS, CXXFLAGS and LDFLAGS for the platform/compiler

2. DCE/RPC supports remote procedure calls through the use of known fixed size types.  Specifically dcerpc/include/dce/platform will include platform specific versions of the NDR (Network Data Representation) types, their data representation and related marshalling macros.  You should be able to use existing include files as the basis for new platform ones. The platform include is selected based on the ${target_cpu} set in the dcerpc/MakeKitBuild file.

3. OpenSSL provides its own Configure script which is called via a MakeKitBuild file; it passes the os/compiler information to OpenSSL Configure. To see the supported os/compile combinations run openssl ./Configure. Modify the openssl MakeKitBuild file configure_openssl() function as required to pass the correct os/compiler for your platform.

You can see what was done to port to s390 and PPC/PPC little endian platforms at these commits:
[s390 port](https://github.com/BeyondTrust/pbis-open/tree/a070c203f8d50a11920388a575feeb6778b2e3ee)
[PPC port](https://github.com/BeyondTrust/pbis-open/tree/9384135006da1a2bf68cae029151fe900cc82db7)

**Comment by aldyleongarcia** [comment]
When compiled for the mips64el architecture on Loongson PC
Proceed:
mkdir debug && cd debug
../configure (args)
make
As a result the binaries .deb.

I install them in the debian system with dpkg -i pbis-open.deb
In the configuration stage of pbis-open the liblinenoise-mob.so.0 library is requested which is not present in the .deb file, but was generated during the make, and exists in the folder ./debug/stage.
should be present in /opt/pbis-open/lib/liblinenoise-mob.so.0 
Can you give me advice to add this library

**Comment by dmorash-BT** [comment]
That is a defect, thanks for pointing that out.  We have a fix for that which we will be pushing to master shortly.  In the interim you can apply this patch to fix your local copy.  You'll need to go through the configure and make package steps again after applying this.
[fix-missing-linenoise-mob.txt](https://github.com/BeyondTrust/pbis-open/files/2676584/fix-missing-linenoise-mob.txt)

**Comment by cboutilier-bt** [comment]
The linenoise-mob library has been added to the package and merged. This has been fixed.

**Comment by luisdanielsc** [comment]
@rbest-bt today I have built, ran and test pbis-open (only CLI) on a Loongson machine with arch MIPS64EL, it worked like supposed. But I can't to build it with support for GUI using the bundled libxml2 becouse libglade2 have many undefined reference to some symbols on that library.

How to build pbis-open using host's available libraries?

**Comment by dmorash-BT** [comment]
I can't speak to the issues with libglade2 and libxml2 right now, but you don't really need the domain join gui.  It is deprecated and very limited in comparison to the CLI domainjoin-cli command, just compare the domainjoin-cli options as documented in the man page with the limited options in the domainjoin-gui.  The domainjoin-cli also offers much better options for debugging why a domainjoin might have failed.

**Comment by luisdanielsc** [comment]
Yes, I know those benefits, but think about the basic users of the system. In that case, it is always a better idea to avoid terminal commands.

**Comment by rbest-bt** [comment]
Merged in #199

pbis-open will longer receive updates and will be archived. Closing all outstanding issues. Please consider BeyondTrust Active Directory Bridge for continued support.  
https://www.beyondtrust.com/privilege-management/active-directory-bridge
